src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+    }
+    else
+    {
+        tts.SetNull();
+        sendMissingTrustedTimeSourceEvent(commandPath.mEndpointId);
+    }
+
+    TimeSynchronizationServer::Instance().SetTrustedTimeSource(tts);
+    commandObj->AddStatus(commandPath, Status::Success);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTimeZoneCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTimeZone::DecodableType & commandData)
+{
+    const auto & timeZone = commandData.timeZone;
+
+    CHIP_ERROR err = TimeSynchronizationServer::Instance().SetTimeZone(timeZone);
+    if (err != CHIP_NO_ERROR)
+    {
+        if (err == CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            commandObj->AddStatus(commandPath, Status::ResourceExhausted);
+        }
+        else if (err == CHIP_IM_GLOBAL_STATUS(InvalidCommand))
+        {
+            commandObj->AddStatus(commandPath, Status::InvalidCommand);
+        }
+        else
+        {
+            commandObj->AddStatus(commandPath, Status::ConstraintError);
+        }
+        return true;
+    }
+
+    if (to_underlying(TimeSynchronizationServer::Instance().GetEventFlag()) & to_underlying(TimeSyncEventFlag::kTimeZoneStatus))
+    {
+        TimeSynchronizationServer::Instance().ClearEventFlag(TimeSyncEventFlag::kTimeZoneStatus);
+        sendTimeZoneStatusEvent(commandPath.mEndpointId);
+    }
+    GetDelegate()->HandleTimeZoneChanged(TimeSynchronizationServer::Instance().GetTimeZone());
+
+    TimeZoneDatabaseEnum tzDb;
+    TimeZoneDatabase::Get(commandPath.mEndpointId, &tzDb);
+    Commands::SetTimeZoneResponse::Type response;
+    TimeSynchronizationServer::Instance().GetUpdatedTimeZoneState();
+    const auto & tzList = TimeSynchronizationServer::Instance().GetTimeZone();
+    if (GetDelegate()->HasFeature(commandPath.mEndpointId, Feature::kTimeZone) && tzDb != TimeZoneDatabaseEnum::kNone &&
+        tzList.size() != 0)
+    {
+        auto & tz = tzList[0].timeZone;
+        if (tz.name.HasValue() && GetDelegate()->HandleUpdateDSTOffset(tz.name.Value()))
+        {
+            response.DSTOffsetRequired = false;
+            sendDSTStatusEvent(commandPath.mEndpointId, true);
+        }
+        else
+        {
+            TimeState dstState = TimeSynchronizationServer::Instance().GetUpdatedDSTOffsetState();
+            TimeSynchronizationServer::Instance().ClearDSTOffset();
+            sendDSTTableEmptyEvent(commandPath.mEndpointId);
+            if (dstState == TimeState::kActive || dstState == TimeState::kChanged)
+            {
+                sendDSTStatusEvent(commandPath.mEndpointId, false);
+            }
+            response.DSTOffsetRequired = true;
+        }
+    }
+    else
+    {
+        response.DSTOffsetRequired = true;
+    }
+    commandObj->AddResponse(commandPath, response);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetDSTOffsetCallback(
```

Comment:
This can end up clearing the list without generating a table-empty event, right?

Followup is fine, but perhaps some of this event generation should move into the TimeSynchronizationServer::Instance() methods, to make it harder to get wrong?

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.